### PR TITLE
Experiment with cleaner types, and pass world down

### PR DESCRIPTION
**Pull request by Drache93**
First look at cleaner extended types:

image

### CODE REVIEW

**Comment by Drache93** [comment]
Document type now gets built from keys in extended type, so no parent data (i.e. nice and clean):

image

File: types/index.d.ts [hunk 1]
```diff
@@ -39,328 +38,335 @@ declare module nlp {
...
+    join(str?: string): Document<W>
 
     // Output
     /**  return the document as text */
     text(options?: string | object): string
     /**  pull out desired metadata from the document */
     json(options?: string | object): any
     /** some named output formats */
-    out(format?: string): string
+    out(format?: 'grid' | 'color' | 'normal' | 'csv' | 'newlines' | 'html' | 'text'): string
```
Comment (Drache93) [inline types/index.d.ts]: @spencermountain Would you mind double checking these? I did it awhile ago...

Comment (spencermountain) [reply types/index.d.ts]: Sure, I'll update this on dev locally.

**Comment by spencermountain** [comment]
Gonna merge in and take a crack at it while Britain is asleep.
THANK YOU 🤗

**Comment by spencermountain** [comment]
hey @Drache93 I've added types for the other plugins, according to your newest format.
I also fixed that emoji-bug, so we can release 12.3.0 once you're ready. Lemme know if you're at a comfortable place with it, or would like another day.
thanks

**Comment by Drache93** [comment]
Anywhere better to talk other than closed PRs? 🤣 

Looks all good, think I'm happy with it for now.
Only thing...

Am I right that the functions in plugins that return documents should return extended documents?...This could be a little weird since they should return document will all extensions, not just there own...
Also the plugins that add functions that return a class, should they return an extended document/basic one/or a reduced one?

iirc/if I understand correctly...the classes introduced in plugins reimplement document functions (i.e. they aren't using e.g. document implementation of forEach, they are implementing their own one)? In which case, they don't need to worry about other plugins add, the functions would just need to correctly return the class

Make sense? :D

**Comment by spencermountain** [comment]
yeah, i understand. It is weird.
`doc.numbers().foo()` should return normal Doc objects, i think.
although I'm really not too sure.
Right now, at least, they do.

the classes created in the plugins are supersets of the `Doc` class - they don't usually re-implement forEach. They only times when they do, is when the method is overloaded, which is rare. `.json()` for example, gets overloaded frequently.

Hey, are you seeing this, too?
image

if I run `ts-node /types/types.test.ts` i get the same error. Is there something I can do?

**Comment by Drache93** [comment]
Oh that's my bad, testing the main deps is now dependant on the numbers one working (so for me that was install and link). Probably not helpful since that's not automatic! Will include the types in the top level test.

**Comment by Drache93** [comment]
So should `doc.numbers().numbers()` work?

**Comment by spencermountain** [comment]
haha - oh boy. Never thought of that.
Yes it does. I just tried it.
But I can live with that ;)
deploying 12.3 to master - 🎈
I would not have been able to do this without your help.

**Comment by Drache93** [comment]
Haha, will sort it out later then. Np! Thanks for the library :D

**Pull request merged**